### ISSUE TITLE
go: binaries built with "buildGoModule" return an empty string for "runtime.GOROOT()"

### ISSUE THREAD

**Issue by leonklingele** [NixOS/nixpkgs#224701]
### Describe the bug

Binaries built using `buildGoModule` return an empty string for `runtime.GOROOT()`. This will change behavior and most likely break all apps relying on `GOROOT`.

I noticed that `swag init` (see [`go-swag`](https://github.com/NixOS/nixpkgs/blob/fb3a9cec932aa6fe44119f944389c7c5a1b02070/pkgs/development/tools/go-swag/default.nix)) was producing inconsistent results based on whether it was installed through `nix-shell --pure -p go-swag`, installed using `go install github.com/swaggo/swag/cmd/swag@latest` or build manually from the GitHub repo.

`go-swag` relies on `runtime.GOROOT()` to ignore files from the Go stdlib: https://github.com/swaggo/swag/blob/e67c96576bc94cbb0acc8026885ed53e0b08ad04/packages.go#L96

This prefix-check always returns true due to `runtime.GOROOT() == ""`, effectively ignoring _all_ files.

As it is unknown if this breaks other apps (due to them depending directly on `GOROOT`, or one of their dependencies), this should be fixed in `buildGoModule` and not only for the `go-swag` nixpkg.

### Steps To Reproduce

Minimal reproducible here: https://git.helsinki.tools/leon.klingele/nix-buildgomodule-bug
Build it using `buildGoModule`. Running it will output something similar to

```sh
[..]
runtime.GOROOT() == "" true
```

### Expected behavior

I don't have a nix system at hand so I can't tell the exact output. But that one line which says `true` in the output above should say `false` instead.

### Notify maintainers

Sorry for the spam. I'm going to mentioning all users who have contributed to the Go compiler package this year.

@dotlambda @zowoq @Stunkymonkey @zhaofengli @trofi @yu-re-ka

### Metadata

s.a., will update tomorrow if required.

**Comment by trofi** [comment]
Let's pull in `golang team`:

`map (x: "@${x.github}") lib.teams.golang.members` expands to: @c00w @kalbasit @Mic92 @zowoq @qbit

**Comment by mtoohey31** [comment]
I believe this is a result of the program being compiled with the `-trimpath` argument. This can be disabled by passing `allowGoReference` to `buildGoModule`. Having `allowGoReference = false` is desirable so that Go programs' closures don't include the Go compiler by default.

**Comment by 06kellyjac** [comment]
> It uses the GOROOT environment variable, if set at process start, or else the root used during the Go build. 

Yep either drop `-trimpath` or provide a `GOROOT` env var either when running the command or use a wrapper.

---

Although in this case I think this a bug that go-swag needs to resolve. It should check if the GOROOT function actually gives a value & not just ignore everything. Then it wouldn't matter if we're using `-trimpath`

**Comment by leonklingele** [comment]
Any news on this?

The same issue just struck us again with `go-licenses` which doesn't work at all due to the use of `runtime.GOROOT()` here: https://github.com/google/go-licenses/blob/150520fc5b3aaf07a90ff6cc50f7227abda448f8/licenses/library.go#L303

**Comment by 06kellyjac** [comment]
What resolution do you expect?

trimpath is default and `allowGoReference` removes trimpath for cases where you need GOROOT or other stuff trimpath removes.
Does that still not resolve the issue for `go-licenses`?
